File: app/framework/qt-transport/tests/web/theme_bridge_test.ts
import { test, expect, beforeAll, afterAll } from 'bun:test'
import { createWsConnection, type BridgeConnection } from '@app/bridge/lib/transport/bridge-transport'
import { spawn, type Subprocess } from 'bun'
import fs from 'fs'

// These tests launch the REAL dev-server and verify that ThemeBridge's
// theme state methods and signal round-trip work through the bridge protocol.
//
// The dev-server self-wires qtThemeRequested → updateQtThemeState so the full
// signal loop is testable without a StyleManager.

const PORT = 19879
let server: Subprocess
let conn: BridgeConnection

interface ThemeBridge {
  setQtTheme(req: { displayName: string; isDark: boolean }): Promise<{ ok: boolean }>
  getQtTheme(): Promise<{ displayName: string; isDark: boolean }>
  getQtThemeFilePath(): Promise<{ path: string; embedded: boolean }>
  qtThemeChanged: (cb: (data: any) => void) => () => void
}

beforeAll(async () => {
  const binaryPath = fs.readFileSync('build/.dev-server-binary.txt', 'utf8').trim()
  if (!fs.existsSync(binaryPath))
    throw new Error(`dev-server binary not found at ${binaryPath} — run xmake build dev-server`)

  server = spawn([binaryPath, '--port', String(PORT)], {
    stdout: 'pipe',
    stderr: 'pipe',
  })

  const start = Date.now()
  while (Date.now() - start < 10000) {
    try {
      conn = await createWsConnection(`ws://localhost:${PORT}`)
      return
    } catch {
      await new Promise(r => setTimeout(r, 200))
    }
  }
  throw new Error('dev-server did not start within 10s')
})

afterAll(() => {
  server?.kill()
})

function bridge() {
  return conn.bridge<ThemeBridge>('theme')
}

// ── setQtTheme ──────────────────────────────────────────────────────

test('setQtTheme returns ok', async () => {
  const result = await bridge().setQtTheme({ displayName: 'Zinc', isDark: true })
  expect(result.ok).toBe(true)
})

// ── getQtTheme ──────────────────────────────────────────────────────

test('getQtTheme returns state set via setQtTheme (self-wired round-trip)', async () => {
  await bridge().setQtTheme({ displayName: 'Catppuccin', isDark: false })
  const result = await bridge().getQtTheme()
  expect(result.displayName).toBe('Catppuccin')
  expect(result.isDark).toBe(false)
})

test('getQtTheme reflects latest state after multiple sets', async () => {
  await bridge().setQtTheme({ displayName: 'Rose Pine', isDark: true })
  await bridge().setQtTheme({ displayName: 'Dracula', isDark: false })
  const result = await bridge().getQtTheme()
  expect(result.displayName).toBe('Dracula')
  expect(result.isDark).toBe(false)
})

// ── getQtThemeFilePath ──────────────────────────────────────────────

test('getQtThemeFilePath returns default empty state', async () => {
  const result = await bridge().getQtThemeFilePath()
  expect(result).toHaveProperty('path')
  expect(result).toHaveProperty('embedded')
})

// ── qtThemeChanged signal ───────────────────────────────────────────

test('qtThemeChanged signal fires when setQtTheme is called', async () => {
  const received = new Promise<any>((resolve) => {
    bridge().qtThemeChanged(resolve)
  })

  await bridge().setQtTheme({ displayName: 'Nord', isDark: true })

  const data = await received
  expect(data.displayName).toBe('Nord')
  expect(data.isDark).toBe(true)
})

File: app/bridges/theme/tests/unit/theme_bridge_test.cpp
#include <catch2/catch_test_macros.hpp>
#include "theme_bridge.hpp"

TEST_CASE("getQtTheme returns stored state after updateQtThemeState") {
    ThemeBridge bridge;
    bridge.updateQtThemeState("Synthwave '84", true);

    auto result = bridge.dispatch("getQtTheme", nlohmann::json::object());
    REQUIRE(result["displayName"] == "Synthwave '84");
    REQUIRE(result["isDark"] == true);
}

TEST_CASE("getQtTheme returns defaults before any update") {
    ThemeBridge bridge;

    auto result = bridge.dispatch("getQtTheme", nlohmann::json::object());
    REQUIRE(result["displayName"] == "");
    REQUIRE(result["isDark"] == true);
}

TEST_CASE("setQtTheme emits qtThemeRequested signal with payload") {
    ThemeBridge bridge;
    nlohmann::json received;
    bridge.on_signal("qtThemeRequested", [&](const nlohmann::json& data) {
        received = data;
    });

    bridge.dispatch("setQtTheme", {{"displayName", "Zinc"}, {"isDark", false}});

    REQUIRE(received["displayName"] == "Zinc");
    REQUIRE(received["isDark"] == false);
}

TEST_CASE("setQtTheme returns ok") {
    ThemeBridge bridge;

    auto result = bridge.dispatch("setQtTheme", {{"displayName", "Zinc"}, {"isDark", true}});
    REQUIRE(result["ok"] == true);
}

TEST_CASE("updateQtThemeState emits qtThemeChanged signal") {
    ThemeBridge bridge;
    nlohmann::json received;
    bridge.on_signal("qtThemeChanged", [&](const nlohmann::json& data) {
        received = data;
    });

    bridge.updateQtThemeState("Catppuccin", false);

    REQUIRE(received["displayName"] == "Catppuccin");
    REQUIRE(received["isDark"] == false);
}

TEST_CASE("getQtThemeFilePath returns stored file path") {
    ThemeBridge bridge;
    bridge.setQtThemeFilePath("/styles/compiled/zinc-dark.qss", false);

    auto result = bridge.dispatch("getQtThemeFilePath", nlohmann::json::object());
    REQUIRE(result["path"] == "/styles/compiled/zinc-dark.qss");
    REQUIRE(result["embedded"] == false);
}

TEST_CASE("getQtThemeFilePath returns embedded flag correctly") {
    ThemeBridge bridge;
    bridge.setQtThemeFilePath(":/styles/default-dark.qss", true);

    auto result = bridge.dispatch("getQtThemeFilePath", nlohmann::json::object());
    REQUIRE(result["path"] == ":/styles/default-dark.qss");
    REQUIRE(result["embedded"] == true);
}

TEST_CASE("unknown method returns error") {
    ThemeBridge bridge;

    auto result = bridge.dispatch("nonexistent", nlohmann::json::object());
    REQUIRE(result.contains("error"));
    REQUIRE(result["error"].get<std::string>().find("Unknown method") != std::string::npos);
}
